ykrt/src/compile/j2/opt/cse.rs

```diff
@@ -0,0 +1,345 @@
+        self.heads[dim_off] = InstIdx::from_usize(self.predecessors.len() - 1);
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::compile::j2::opt::{
+        opt::{OptOutcome, test::opt_and_test},
+        strength_fold::strength_fold,
+    };
+    use std::{cell::RefCell, rc::Rc};
+
+    fn test_cse(mod_s: &str, ptn: &str) {
+        let cse = Rc::new(RefCell::new(CSE::new()));
+        opt_and_test(
+            mod_s,
+            |opt, mut inst| {
+                if let Inst::Guard(_) = inst {
+                    match strength_fold(opt, inst) {
+                        OptOutcome::NotNeeded => OptOutcome::NotNeeded,
+                        OptOutcome::Rewritten(inst) => {
+                            let mut cse = cse.borrow_mut();
+                            cse.push(&inst);
+                            OptOutcome::Rewritten(inst)
+                        }
+                        OptOutcome::Equiv(iidx) => OptOutcome::Equiv(iidx),
+                    }
+                } else {
+                    inst.canonicalise(opt);
+                    let mut cse = cse.borrow_mut();
+                    if let Some(iidx) = cse.is_equiv(opt, &inst) {
+                        OptOutcome::Equiv(iidx)
+                    } else {
+                        cse.push(&inst);
+                        OptOutcome::Rewritten(inst)
+                    }
+                }
+            },
+            ptn,
+        );
+    }
+
+    #[test]
+    fn cse() {
+        test_cse(
+            "
+          %0: i8 = arg [reg]
+          %1: i8 = arg [reg]
+          %2: i8 = add %0, %1
+          %3: i8 = add %0, %1
+          blackbox %2
+          blackbox %3
+        ",
+            "
+          %0: i8 = arg
+          %1: i8 = arg
+          %2: i8 = add %0, %1
+          blackbox %2
+          blackbox %2
+        ",
+        );
+    }
+
+    #[test]
+    fn cse_equiv() {
+        // Test that deep instruction equivalence works
+        test_cse(
+            "
+          %0: i8 = arg [reg]
+          %1: i8 = arg [reg]
+          %2: i1 = icmp eq %0, %1
+          guard true, %2, []
+          %4: i8 = add %0, %0
+          %5: i8 = add %1, %1
+          blackbox %4
+          blackbox %5
+        ",
+            "
+          %0: i8 = arg
+          %1: i8 = arg
+          %2: i1 = icmp eq %0, %1
+          guard true, %2, []
+          %4: i8 = add %0, %0
+          blackbox %4
+          blackbox %4
+        ",
+        );
+
+        test_cse(
+            "
+          %0: i8 = arg [reg]
+          %1: i8 = arg [reg]
+          %2: i8 = add %0, %0
+          %3: i1 = icmp eq %0, %1
+          guard true, %3, []
+          %5: i8 = add %1, %1
+          blackbox %2
+          blackbox %5
+        ",
+            "
+          %0: i8 = arg
+          %1: i8 = arg
+          %2: i8 = add %0, %0
+          %3: i1 = icmp eq %0, %1
+          guard true, %3, []
+          blackbox %2
+          blackbox %2
+        ",
+        );
```

Comment:
Can you please add one more test where CSE is done over a side-effecting instruction as well?
Maybe
```
          %0: ptr = arg [reg]
          %1: ptr = arg [reg]
          %2: i64 = arg [reg]
          %3: i64 = add %0, %1
          memcpy %0, %1, %2, true
          %4: i64 = add %0, %1 // This should be CSE-d as it is equivalent to %3 trivially.
          blackbox ...
```

Reply:
Test added in 8be0308e.